### PR TITLE
Proposed refactor to better encapsulate reconciliation

### PR DESCRIPTION
**Pull request by ellis-bigelow**
I think we can better encapsulate the subresource logic by moving the resource creation into the individual reconcilers.

Each reconciler may modify the status of the passed in KFService. The top level ReconcileService is responsible for finally updating the changes.

WDYT?

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/147)

### CODE REVIEW

**Comment by rakelkar** [comment]
I think this is a great idea and will make everything way more individually testable..!

File: pkg/reconciler/knative/resources/knative_configuration.go [hunk 1]
```diff
@@ -34,26 +38,26 @@ const (
 )
 
 type ConfigurationBuilder struct {
-	frameworksConfig *v1alpha1.FrameworksConfig
+	frameworksConfig  *v1alpha1.FrameworksConfig
+	credentialBuilder *credentials.CredentialBuilder
 }
 
-func NewConfigurationBuilder(config *v1.ConfigMap) *ConfigurationBuilder {
+func NewConfigurationBuilder(client client.Client, config *v1.ConfigMap) ConfigurationBuilder {
```
Comment (ellis-bigelow) [inline pkg/reconciler/knative/resources/knative_configuration.go]: What's the best practice here for returning *ConfigurationBuilder vs ConfigurationBuilder

**Comment by ellis-bigelow** [comment]
Hey all. I finished the implementation of the new reconcilers. I like the code pattern, and I think it separates concerns pretty well. Our object generation is strongly decoupled from the object creation logic, which I think makes the code base cleaner.

I have some thoughts on the testing. So far, we've been testing without mocks, which means that there's a ton of test scaffolding necessary for code at each level. The scaffolding is different if we're doing

1. resource generation
2. resource reconciliation
3. top level reconciliiation

We end up testing the main code paths many many times to cover the edge cases.

I think we need to pick one direction and go that way.

## 1, Mock Interfaces
One approach in many languages is to mock all interfaces and only test the behavior of the file under test. This is nice and clean, but doesn't cover e2e integration.

## 2, Run all tests as e2e
Another approach is to have strong e2e tests that cover all code paths. The caveat here is that all e2e tests run from the same level, so you can hide most of the setup/teardown boilerplate into a helper function, which makes writing individual tests easier.

File: pkg/controller/kfservice/kfservice_controller.go [hunk 2]
```diff
@@ -117,6 +110,11 @@ type ReconcileService struct {
 	Recorder record.EventRecorder
 }
 
+// Reconciler is implemented by all subresources
+type Reconciler interface {
```
Comment (rakelkar) [inline pkg/controller/kfservice/kfservice_controller.go]: The controller runtime also defines an interface named Reconciler, maybe rename this SubReconciler or something?

**Comment by rakelkar** [comment]
I think the biggest benefit of this pattern is that it enables clean separation and mocking. 

I suggest we have:
* one e2e test that tests a simple golden path for the main Reconcile method (maybe just create?)
* all other tests on Reconcile based on mocks (if it has any logic beyond just calling the sub-reconcilers). * individual sub-reconcilers can now have targeted tests.

**Comment by yuzisun** [comment]
I like the refactoring! 
For testing I am not clear what's the mocking we are talking about here, we are currently using https://godoc.org/sigs.k8s.io/controller-runtime/pkg/envtest with local mock kubernetes controller panel for apiserver and etcd. If we are referring to mocking the kubernetes client side `Create` and `Update` API etc, I have done this before it does not work well because it misses tons of kubernetes test cases e.g if you set an invalid kubernetes resource name your test still passes which then gives a lot burden on writing comprehensive e2e tests in a real environment. The downside of full e2e tests is that it takes long time to setup environment and run all the test cases, from my experience 80% test cases should be covered by unit/integration tests and 20% covered by e2e. I kind of like the idea of `controller-runtime` to use the local controller panel, if you have concern about testing maybe we can chat with kube builder/controller-runtime community ? 
On the other note that we are testing main path many times, I think that's mainly because we have different levels for resource builder, subresource reconciliation and top level reconciliation,  the scenarios might be duplicated but tests are still necessary. There might be better way but I am not clear how mock or e2e solves the problem.

**Comment by ellis-bigelow** [comment]
> I think the biggest benefit of this pattern is that it enables clean separation and mocking.
> 
> I suggest we have:
> 
> * one e2e test that tests a simple golden path for the main Reconcile method (maybe just create?)
> * all other tests on Reconcile based on mocks (if it has any logic beyond just calling the sub-reconcilers). * individual sub-reconcilers can now have targeted tests.

This is the approach I was leaning as well :)

**Comment by ellis-bigelow** [comment]
Ok this is ready.

**Comment by yuzisun** [comment]
> Ok this is ready.

Awesome! I will review this late tonight.

File: pkg/controller/kfservice/reconcilers/knative/configuration_reconciler.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var log = logf.Log.WithName("Reconciler")
```
Comment (yuzisun) [inline pkg/controller/kfservice/reconcilers/knative/configuration_reconciler.go]: Are we going to migrate all to klog?

File: pkg/controller/kfservice/kfservice_controller.go (hunk 2 above)
Comment (ellis-bigelow) [reply pkg/controller/kfservice/kfservice_controller.go]: I honestly think this is fine as is. It's the same concept.

File: pkg/controller/kfservice/reconcilers/knative/configuration_reconciler.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	}
+
+	status, err := r.reconcileConfiguration(kfsvc, &defaultConfiguration)
+	if err != nil {
+		return err
+	}
+
+	kfsvc.Status.PropagateDefaultConfigurationStatus(&status)
+
+	if kfsvc.Spec.Canary == nil {
```
[the remark this answers is not in the document]
Comment (ellis-bigelow) [reply pkg/controller/kfservice/reconcilers/knative/configuration_reconciler.go]: dead code. removed

File: pkg/controller/kfservice/reconcilers/knative/configuration_reconciler.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	if err := r.reconcileDefault(kfsvc); err != nil {
+		return err
+	}
+	if err := r.reconcileCanary(kfsvc); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *ConfigurationReconciler) reconcileDefault(kfsvc *v1alpha1.KFService) error {
```
[the remark this answers is not in the document]
Comment (ellis-bigelow) [reply pkg/controller/kfservice/reconcilers/knative/configuration_reconciler.go]: I tried to do this, but the differences weren't enough to really clean up the code. I was thinking about putting the controller references into the configuration builder, but that broke the encapsulation of that interface. I pushed that piece into shared, which makes it a bit shorter.

File: pkg/controller/kfservice/resources/knative/configuration.go [hunk 6]
```diff
@@ -14,44 +14,50 @@ See the License for the specific language governing permissions and
...
+
 	return &ConfigurationBuilder{
-		frameworksConfig: frameworkConfig,
+		frameworksConfig:  frameworkConfig,
+		credentialBuilder: credentials.NewCredentialBulder(client, config),
 	}
 }
 
-func (c *ConfigurationBuilder) CreateKnativeConfiguration(name string, metadata metav1.ObjectMeta, modelSpec *v1alpha1.ModelSpec) *knservingv1alpha1.Configuration {
+func (c *ConfigurationBuilder) CreateKnativeConfiguration(name string, metadata metav1.ObjectMeta, modelSpec *v1alpha1.ModelSpec) (knservingv1alpha1.Configuration, error) {
```
[the remark this answers is not in the document]
Comment (ellis-bigelow) [reply pkg/controller/kfservice/resources/knative/configuration.go]: Sure.

File: pkg/controller/kfservice/reconcilers/knative/configuration_reconciler.go (hunk 3 above)
Comment (ellis-bigelow) [reply pkg/controller/kfservice/reconcilers/knative/configuration_reconciler.go]: Yes but not in this PR. Not sure why these weren't before.

File: pkg/controller/kfservice/reconcilers/knative/configuration_reconciler.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (yuzisun) [reply pkg/controller/kfservice/reconcilers/knative/configuration_reconciler.go]: did you push the change? looks still there
